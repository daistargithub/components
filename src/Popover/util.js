export const POSITIONS = {
  TOP: 'top',
  TOP_LEFT: 'top-left',
  TOP_RIGHT: 'top-right',
  BOTTOM: 'bottom',
  BOTTOM_LEFT: 'bottom-left',
  BOTTOM_RIGHT: 'bottom-right',
  LEFT: 'left',
  RIGHT: 'right'
}

export function getPositions (props) {
  let bottom = 'auto'
  let left = 'auto'
  let right = 'auto'
  let top = 'auto'
  let transform = 'translate(0, 0)'

  switch (props.position) {
    case POSITIONS.TOP:
      bottom = '100%'
      left = '50%'
      transform = 'translate(-50%, 0)'
      break

    case POSITIONS.BOTTOM:
      top = '100%'
      left = '50%'
      transform = 'translate(-50%, 0)'
      break

    case POSITIONS.LEFT:
      right = '100%'
      top = '50%'
      transform = 'translate(0, -50%)'
      break

    case POSITIONS.RIGHT:
      left = '100%'
      top = '50%'
      transform = 'translate(0, -50%)'
      break

    case POSITIONS.TOP_LEFT:
      bottom = '100%'
      left = 0
      break

    case POSITIONS.BOTTOM_LEFT:
      top = '100%'
      left = 0
      break

    case POSITIONS.TOP_RIGHT:
      bottom = '100%'
      right = 0
      break

    case POSITIONS.BOTTOM_RIGHT:
      top = '100%'
      right = 0
      break
  }

  return {
    bottom,
    left,
    right,
    top,
    transform
  }
}
