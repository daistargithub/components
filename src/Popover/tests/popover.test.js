import 'jest-styled-components'
import React from 'react'
import { containerRenderer as renderer } from 'utils/tests/containerRenderer'
import { Popover } from '../index'
import { POSITIONS } from '../util'

// NOTE: since we are not using enzyme or react-testing-library we
// are not simulating / testing the click or esc key events

test('Popover renders', () => {
  const component = renderer
    .create(
      <Popover
        isShown
        render={() => (
          <div>
            <p>Popover content</p>
          </div>
        )}
      >
        <button>Regular Popover</button>
      </Popover>
    )
    .toJSON()
  expect(component).toMatchSnapshot()
})

test('Supports default visibility via isShown prop', () => {
  const component = renderer
    .create(
      <Popover
        isShown
        render={() => (
          <div>
            <p>Popover content</p>
          </div>
        )}
      >
        <button>Regular Popover</button>
      </Popover>
    )
    .toJSON()
  expect(component).toMatchSnapshot()
})

test('Supports function as a child', () => {
  const component = renderer
    .create(
      <Popover
        isShown
        render={() => (
          <div>
            <p>Popover content</p>
          </div>
        )}
      >
        {({ ref, toggle }) => (
          <button ref={ref} onClick={toggle}>
            Regular Popover
          </button>
        )}
      </Popover>
    )
    .toJSON()
  expect(component).toMatchSnapshot()
})

test('Supports nested Popovers', () => {
  const component = renderer
    .create(
      <Popover
        isShown
        render={() => (
          <div>
            <p>Popover content</p>
            <Popover
              isShown
              render={() => (
                <div>
                  <p>Popover content</p>
                </div>
              )}
            >
              <button>Regular Popover</button>
            </Popover>
          </div>
        )}
      >
        <button>Nested Popover</button>
      </Popover>
    )
    .toJSON()

  expect(component).toMatchSnapshot()
})

test('Supports different positions', () => {
  const positions = Object.values(POSITIONS)

  positions.forEach(pos => {
    const component = renderer
      .create(
        <Popover
          isShown
          position={pos}
          render={() => (
            <div>
              <p>Popover content</p>
            </div>
          )}
        >
          <button>Regular Popover</button>
        </Popover>
      )
      .toJSON()

    // NOTE: to test inner content like this we would need to use something
    // like enzyme or react-testing-library as jest-styled-components only
    // checks the root element of a rendered tree.
    // REF: https://github.com/styled-components/jest-styled-components#tohavestylerule

    // const positionStyles = getPositions({ position: pos })
    // const positionStyleKeys = Object.keys(positionStyles)

    // positionStyleKeys.forEach(styleProp => {
    //   expect(component).toHaveStyleRule(styleProp, positionStyles[styleProp])
    // })

    expect(component).toMatchSnapshot()
  })
})
