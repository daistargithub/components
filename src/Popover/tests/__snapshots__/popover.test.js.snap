// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover renders 1`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: 0;
  right: auto;
  top: 100%;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports default visibility via isShown prop 1`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: 0;
  right: auto;
  top: 100%;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports different positions 1`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: 100%;
  left: 50%;
  right: auto;
  top: auto;
  -webkit-transform: translate(-50%,0);
  -ms-transform: translate(-50%,0);
  transform: translate(-50%,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports different positions 2`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: 100%;
  left: 0;
  right: auto;
  top: auto;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports different positions 3`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: 100%;
  left: auto;
  right: 0;
  top: auto;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports different positions 4`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: 50%;
  right: auto;
  top: 100%;
  -webkit-transform: translate(-50%,0);
  -ms-transform: translate(-50%,0);
  transform: translate(-50%,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports different positions 5`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: 0;
  right: auto;
  top: 100%;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports different positions 6`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: auto;
  right: 0;
  top: 100%;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports different positions 7`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: auto;
  right: 100%;
  top: 50%;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports different positions 8`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: 100%;
  right: auto;
  top: 50%;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports function as a child 1`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: 0;
  right: auto;
  top: 100%;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Regular Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
    </div>
  </div>
</div>
`;

exports[`Supports nested Popovers 1`] = `
.c0 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  max-width: 100%;
  position: relative;
}

.c1 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 10;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: 0;
  right: auto;
  top: 100%;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

.c2 {
  box-sizing: border-box;
  box-shadow: 0 0 0 rgba(0,0,0,0.12);
  z-index: 11;
  max-width: 100%;
  position: absolute;
  bottom: auto;
  left: 0;
  right: auto;
  top: 100%;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
}

<div
  className="c0"
  elevation={0}
>
  <button
    onClick={[Function]}
  >
    Nested Popover
  </button>
  <div
    className="c1"
    elevation={0}
  >
    <div>
      <p>
        Popover content
      </p>
      <div
        className="c0"
        elevation={0}
      >
        <button
          onClick={[Function]}
        >
          Regular Popover
        </button>
        <div
          className="c2"
          elevation={0}
        >
          <div>
            <p>
              Popover content
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
